### PR TITLE
os:cmd instead of Erlang ports

### LINKED ISSUE #16: Multiple inline commands don't run
If I have `echo a; echo b` only the first echo runs.

### LINKED ISSUE #21: Make command timeout configurable

### LINKED ISSUE #22: Use os:cmd instead of Erlang ports
os:cmd handles many edge cases and in general is better.

### PR DESCRIPTION
**Pull request by paulo-ferraz-oliveira**
Attempting to [Fix #16] [Fix #21] [Fix #22], by replacing Erlang ports with `os:cmd` as discussed in two of the linked Issues.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+language: erlang
```
Comment (paulo-ferraz-oliveira) [inline .travis.yml:1]: I believe you have to allow Travis CI to "read" this repo. in order for tests to run.

File: rebar.config [hunk 2]
```diff
@@ -1,9 +1,32 @@
 {erl_opts, [debug_info]}.
-{plugins, [rebar3_hex, erlfmt]}.
+{plugins, [rebar3_hex, erlfmt, rebar3_lint]}.
```
Comment (paulo-ferraz-oliveira) [inline rebar.config]: For extra juiciness. If you prefer, though, I can remove it.

File: src/rebar_cmd.app.src [hunk 3]
```diff
@@ -1,8 +1,8 @@
 {application, rebar_cmd,
- [{description, "Run custom shell commands as rebar3 tasks"},
-  {vsn, "0.2.6"},
+ [{description, "Run custom shell commands with rebar3 cmd <command>"},
+  {vsn, git},
   {registered, []},
-  {application,
```
Comment (paulo-ferraz-oliveira) [inline src/rebar_cmd.app.src:5]: This was wrong.

File: src/rebar_cmd.app.src [hunk 4]
```diff
@@ -1,8 +1,8 @@
 {application, rebar_cmd,
- [{description, "Run custom shell commands as rebar3 tasks"},
-  {vsn, "0.2.6"},
+ [{description, "Run custom shell commands with rebar3 cmd <command>"},
+  {vsn, git},
```
Comment (paulo-ferraz-oliveira) [inline src/rebar_cmd.app.src]: I can revert this, but this stops you from having to manually remember this at every release.

File: src/rebar_cmd.app.src [hunk 5]
```diff
@@ -1,8 +1,8 @@
 {application, rebar_cmd,
- [{description, "Run custom shell commands as rebar3 tasks"},
-  {vsn, "0.2.6"},
+ [{description, "Run custom shell commands with rebar3 cmd <command>"},
```
Comment (paulo-ferraz-oliveira) [inline src/rebar_cmd.app.src:2]: Do you like it like this? (I also updated the `.erl`); might be reverted if you don't.

File: src/rebar_cmd.app.src [hunk 6]
```diff
@@ -11,5 +11,5 @@
 
   {maintainers, ["Sasan Hezarkhani", "Matheus Mendes", "Ahmad Mokhtar"]},
   {licenses, ["MIT"]},
-  {links, [{"Github", "https://github.com/shezarkhani/rebar_cmd"}]}
+  {links, [{"Github", "https://github.com/gootik/rebar_cmd"}]}
```
Comment (paulo-ferraz-oliveira) [inline src/rebar_cmd.app.src:14]: This redirect is actually what GitHub does already.

File: src/rebar_cmd_prv.erl [hunk 7]
```diff
@@ -30,17 +49,8 @@ init(State) ->
...
-        {_, Command} ->
-            rebar_api:debug("Running ~p with command ~p.~n", [[CmdName], [Command]]),
-            case run_shell({CmdName, Command}) of
-                {ok, _} -> {ok, State};
-                {error, Error} -> {error, Error}
-            end;
-        false ->
-            {error, "Command " ++ CmdName ++ " is unknown. Check your configuration."}
-    end.
+    CmdName = proplists:get_value(task, ArgsPropList, undefined),
```
Comment (paulo-ferraz-oliveira) [inline src/rebar_cmd_prv.erl:57]: `:lookup` was crashing if you did `rebar3 cmd` (without the `<command>`).

File: src/rebar_cmd_prv.erl [hunk 8]
```diff
@@ -49,36 +59,78 @@ format_error(Reason) ->
...
+    {MonitorPid, MonitorRef} = erlang:spawn_monitor(
+        fun() ->
+            not is_list(Cmd) andalso exit("call not expressed as a string"),
+            rebar_api:debug("Command ~s executing ~s with options ~p.", [CmdName, Cmd, Opts]),
+            Pid ! {Ref, os:cmd(Cmd)}
+        end
+    ),
     receive
-        {Port, {data, MoreData}} ->
-            io:format(MoreData),
```
Comment (paulo-ferraz-oliveira) [inline src/rebar_cmd_prv.erl:62]: I'm not sure this was intended or not, but as-is, I'm using the `rebar3` API interface.

File: src/rebar_cmd_prv.erl [hunk 9]
```diff
@@ -49,36 +59,78 @@ format_error(Reason) ->
 %% ===================================================================
 %% Private API
 %% ===================================================================
-run_shell({CmdName, Command}) ->
-    run_shell({CmdName, Command}, 15000).
 
-run_shell({CmdName, Command}, Timeout) ->
-    Port = erlang:open_port({spawn, Command}, [exit_status]),
```
Comment (paulo-ferraz-oliveira) [inline src/rebar_cmd_prv.erl:56]: As discussed elsewhere, you remove the ability to obtain the exit status. Might not be a problem, though. You tell me.

File: .travis.yml (hunk 1 above)
Comment (gootik) [reply .travis.yml:1]: done. thank you

File: rebar.config (hunk 2 above)
Comment (gootik) [reply rebar.config]: i like it :)

File: src/rebar_cmd.app.src (hunk 5 above)
Comment (gootik) [reply src/rebar_cmd.app.src:2]: 👍

File: src/rebar_cmd.app.src (hunk 3 above)
Comment (gootik) [reply src/rebar_cmd.app.src:5]: ooops

File: src/rebar_cmd.app.src (hunk 4 above)
Comment (gootik) [reply src/rebar_cmd.app.src]: yeah let's revert this one, I think it's easier to see what version is running from the file.

File: src/rebar_cmd.app.src (hunk 6 above)
Comment (gootik) [reply src/rebar_cmd.app.src:14]: nice. thank you

File: src/rebar_cmd_prv.erl (hunk 9 above)
Comment (gootik) [reply src/rebar_cmd_prv.erl:56]: Yeah I think this is fine, as long as we can report a failure or a success.

File: src/rebar_cmd_prv.erl [hunk 10]
```diff
@@ -49,36 +59,78 @@ format_error(Reason) ->
...
-                integer_to_list(Timeout) ++ " ms."}
+        exit(MonitorPid, kill),
+        TimeOutS = integer_to_list(Timeout),
+        {error, "Command " ++ CmdName ++ " (" ++ Cmd ++ ") timed out after " ++ TimeOutS ++ " ms."}
     end.
+
+handle_cmd({error, Error}, _Opts, _State) ->
+    {error, Error};
+handle_cmd({ok, {_CmdName, died}}, _Opts, State) ->
+    {ok, State};
```
Comment (gootik) [inline src/rebar_cmd_prv.erl:117]: should we move the error reporting to handle_cmd? seems like a better place

File: src/rebar_cmd_prv.erl [hunk 11]
```diff
@@ -49,36 +59,78 @@ format_error(Reason) ->
...
+    {atom_to_list(CmdName), Cmd, Opts}.
 
-shell_loop(Port, Data, Timeout, {CmdName, Command}) ->
+exec(CmdName, Cmd, Opts) ->
+    Pid = self(),
+    Ref = erlang:make_ref(),
+    Timeout = get_opt(timeout, Opts, CmdName),
+    {MonitorPid, MonitorRef} = erlang:spawn_monitor(
+        fun() ->
+            not is_list(Cmd) andalso exit("call not expressed as a string"),
```
Comment (gootik) [inline src/rebar_cmd_prv.erl:97]: interesting. I like this style

**Review by gootik — COMMENTED** [review]
Looks great. Thank you for all the cleanup you did. This was the first ever thing I made in Erlang and it was due for some love.

File: src/rebar_cmd_prv.erl (hunk 11 above)
Comment (paulo-ferraz-oliveira) [reply src/rebar_cmd_prv.erl:97]: The `andalso` shortcut? It's OK for simple things like this one, to not have a new function or a case/if, but don't go using it too much. It's a deep rabbit whole.

File: src/rebar_cmd_prv.erl (hunk 10 above)
Comment (paulo-ferraz-oliveira) [reply src/rebar_cmd_prv.erl:117]: Do you mean the `rebar_api:error` in `:exec/3`?

Comment (paulo-ferraz-oliveira) [reply src/rebar_cmd_prv.erl:117]: 20dd102

**Comment by paulo-ferraz-oliveira** [comment]
I pushed a fix for the tests. I also ended up having `erlfmt` inside its own profile, because it'll only run in Erlang/OTP 21+ and it was thus restricting the "plugin" which can support from 19.3 on (the same versions `rebar3` supports).

**Comment by paulo-ferraz-oliveira** [comment]
Hurrah, tests passed.
Don't merge just yet. I'm tweaking the `README.md`.

**Edit**: you can see how it looks [here](https://github.com/paulo-ferraz-oliveira/rebar_cmd/blob/feature/os_cmd/README.md).

**Comment by gootik** [comment]
nicely done. I'll merge this now and anything that you like to improve on can be separate PRs.
Again, thank you so much for doing all of this. really appreciate it

**Comment by paulo-ferraz-oliveira** [comment]
If you feel it's OK, could you release it and possibly publish it to Hex? Thanks.

**Pull request merged**